separate tutorials into each document and add images (#1230)
vdaas/vald: docs/tutorial/get-started.md

Vald is a highly scalable distributed fast approximate nearest neighbor dense vector search engine.<br>
Vald is designed and implemented based on Cloud-Native architecture.

This article will show you how to deploy and run the Vald components on your Kubernetes cluster.
And, Fashion-mnist is used as an example of a dataset.

## Overview

Before starting, let's check the below image.
It shows the architecture image about the deployment result of Get Started.<br>
The 4 kinds of components, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will be deployed to the Kubernetes.

<img src="../../assets/docs/tutorial/getstarted.png" />

The 5 steps to Get Started with Vald:
1. [Check and Satisfy the Requirements](#Requirements)
1. [Prepare Kubernetes Cluster](#Prepare-the-Kubernetes-Cluster)
1. [Deploy Vald on Kubernetes Cluster](#Deploy-Vald-on-Kubernetes-Cluster)

> 1. [Run Exapmle Code](#Run-Exapmle-Code)

Reviewer: ```suggestion
1. [Run Example Code](#Run-Example-Code)
```

Vald is designed and implemented based on Cloud-Native architecture.

> This article will show you how to deploy and run the Vald components on your kubernetes cluster.

Reviewer: how about this comment?

> This article will show you how to deploy and run the Vald components on your kubernetes cluster.

Author: I'd like to know why you thought that./

> This article will show you how to deploy and run the Vald components on your kubernetes cluster.

Reviewer: I think this line and the next line has completely different topic, so I think it is better to put a next line character here

> This article will show you how to deploy and run the Vald components on your kubernetes cluster.

Reviewer: @vankichi

> This article will show you how to deploy and run the Vald components on your kubernetes cluster.

Author: > I think this line and the next line has completely different topic, so I think it is better to put a line break here

I don't think so.
Both sentences are describe what user do in this tutorial

And, Fashion-mnist is used as an example of a dataset.

## Overview

Before starting, let's check the below image.
It shows the architecture image about the deployment result of Get Started.<br>

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Author: @kevindiu 
IMO, `Four Vald components` may misread to user.
And for the document, use numeral is better to read. (also IMO)

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Reviewer: ```suggestion
The following Vald components, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager`, will be deployed to the Kubernetes.
```

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Reviewer: BTW, I think `Vald components` is easer to understand.

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Reviewer: Also please at least fix this typo.
```suggestion
The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on Kubernetes.
```

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Author: @kevindiu 
> BTW, I think Vald components is easer to understand.

YAP. For considering only SWE, it is right.

But, this document is not only for SWE.
And we cannot know the user already read other documents like about or architecture.
I think `microservice` is more popular than `components`. how do you think?

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Reviewer 2: @vankichi 
> I think microservice is more popular than components

I think so. in this case, I think microservice is better.

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Reviewer: Consider that, I think component is better, if user do not understand microservice, this word may confuse them. `Component` is more specific than microservice in my point of view.
What do you think? If you still think microservice is better, it is ok for me.

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Author: @kevindiu thx.
your suggestion is good. but, I think using `Components` is difficult (the below is why I thought)

this is the reason why I thought that.
As you know `Vald Agent` is a component, but `vald-agent-ngt-0` is also a component.
On the other hand, it is easy when we use `microservice` for `Vald Agent` and  `component` for `vald-agent-ngt-0`.
if it is not correct, I will change it along with your suggestion. :)

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Reviewer: @vankichi thank you for your comment. 
I think `vald-agent-ngt-0` is the instance of a `Vald Agent` component. 
Microservice means the architecture of the system. and component means the part of the system. so I think component is better.

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Author: instance? 🤔

> The 4 kinds of microservices, `Vald LB Gateway`, `Vald Discoverer`, `Vald Agent` and `Vald Index Manager` will deployed on kuberntes.

Reviewer: yes, the actual instance of a component.

> <img src="../../assets/docs/tutorial/getstarted.png" />

Reviewer: Should we add some brief explanation for these components? 🤔

Author: IMO, it is not the responsibility of this document.
I think it is better to link any document when we create

The 5 steps to Get Started with Vald:
1. [Check and satisfy the requirements](#Requirements)

> 1. [Prepare kubernetes cluster](#Prepare-the-kubernetes-cluster)

Reviewer: ```suggestion
1. [Prepare Kubernetes cluster](#Prepare-the-kubernetes-cluster)
```

> 1. [Deploy Vald on kubernetes cluster](#Deploy-Vald-on-Kubernetes-Cluster)

Reviewer: ```suggestion
1. [Deploy Vald on Kubernetes cluster](#Deploy-Vald-on-Kubernetes-Cluster)
```

> 1. [Running exapmle code](#Running-Exapmle-Code)

Reviewer: the link is dead here 🤔
```suggestion
1. [Run example code](#Run-Example-Code)
```

1. [Cleanup](#Cleanup)

## Requirements

> - kubernetes: v1.19 ~

Reviewer: Please correct all `kubernetes` to `Kubernetes`
```suggestion
- Kubernetes: v1.19 ~
```

- go: v1.15 ~
- helm: v3 ~
- libhdf5 (_only required for get started_)

> Helm is used to deploying Vald on your kubernetes and Hdf5 is used to decode the sample data file to run the example.<br>

Reviewer: ```suggestion
Helm is used to deploying Vald on your Kubernetes and Hdf5 is used to decode the sample data file to run the example.<br>
```

If helm or hdf5 is not installed, please install [helm](https://helm.sh/docs/intro/install) and [hdf5](https://www.hdfgroup.org/).

> <details><summary>Installatoin command for helm</summary><br>

Reviewer: ```suggestion
<details><summary>Installation command for helm</summary><br>
```

    After creating your Kubernetes cluster, let's apply Kubernetes metrics server.

    ```bash
    kubectl apply -f https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml
    kubectl wait -n kube-system --for=condition=ready pod -l k8s-app=metrics-server --timeout=600s
    ```

## Deploy Vald on Kubernetes Cluster

This chapter shows the way to deploy Vald using Helm and to run on your Kubernetes cluster.<br>
In this tutorial, you will deploy the basic configuration of Vald that is consisted of vald-agent-ngt, vald-lb-gateway, vald-discoverer and vald-manager-index.<br>

1. Clone the repository

    To use the `deployment yaml` for deploy, let's clone [`vdaas/vald`](https://github.com/vdaas/vald.git) repository.

    ```bash
    git clone https://github.com/vdaas/vald.git
    cd vald
    ```

1. Confirm which cluster to deploy

   ```bash
   kubectl cluster-info
   ```

1. Deploy Vald using Helm

    ```bash
    # add vald repo into helm repo
    helm repo add vald https://vald.vdaas.org/charts
    # deploy vald on your kubernetes cluster
    helm install vald vald/vald --values example/helm/values.yaml
    ```

1. Verify

>     When finish deploy, you can check the Vald's pods status following command.

Reviewer: ```suggestion
    When finish deploying Vald, you can check the Vald's pods status using the following command.
```

    ```bash
    kubectl get pods
    ```
    <details><summary>Example output</summary><br>
    If the deployment is successful, all Vald components should be running.

    ```bash
    NAME                                       READY   STATUS      RESTARTS   AGE
    vald-agent-ngt-0                           1/1     Running     0          7m12s
    vald-agent-ngt-1                           1/1     Running     0          7m12s
    vald-agent-ngt-2                           1/1     Running     0          7m12s
    vald-agent-ngt-3                           1/1     Running     0          7m12s
    vald-agent-ngt-4                           1/1     Running     0          7m12s
    vald-discoverer-7f9f697dbb-q44qh           1/1     Running     0          7m11s
    vald-lb-gateway-6b7b9f6948-4z5md           1/1     Running     0          7m12s
    vald-lb-gateway-6b7b9f6948-68g94           1/1     Running     0          6m56s
    vald-lb-gateway-6b7b9f6948-cvspq           1/1     Running     0          6m56s
    vald-manager-index-74c7b5ddd6-jrnlw        1/1     Running     0          7m12s
    ```

> ## Run Exapmle Code

Reviewer: ```suggestion
## Run Example Code
```

In this chapter, you will execute insert vectors, search vectors, and delete vectors to your Vald cluster using the example code.<br>
The [fashion-mnist](https://github.com/zalandoresearch/fashion-mnist) is used as a dataset for indexing and search query.

The example code is implemented Go and using [vald-client-go](https://github.com/vdaas/vald-client-go), one of the official Vald client libraries,  for requesting to Vald cluster.
Vald provides multiple language client libraries such as Go, Java, Node.js, Python, and so on.

> If you are interested in, please refer to [SDKs](../user-guides/sdks).<br>

Reviewer: Sorry I made mistake...
```suggestion
If you are interested in, please refer to [SDKs](../user-guides/sdks.md).<br>
```

> If you are interested in, please refer to [SDKs](../../user-guides/sdks).<br>

Reviewer: seems the link is dead..

Reviewer: ```suggestion
If you are interested in, please refer to [SDKs](../user-guides/sdks).<br>
```

1. Port Forward

>     At first, port-foward is required to make request from your local environment possible.

Reviewer: ```suggestion
    At first, port-forward is required to make request from your local environment possible.
```

    ```bash
    kubectl port-forward deployment/vald-lb-gateway 8081:8081
    ```

1. Download dataset

>     Download [fashion-mnist](https://github.com/zalandoresearch/fashion-mnist) that is used as dataset for indexing and search query.

Reviewer: ```suggestion
    Download [fashion-mnist](https://github.com/zalandoresearch/fashion-mnist) that is used as a dataset for indexing and search query.
```

                        Epsilon: 0.01,
                    }
                }
                if err != nil {
                    glg.Fatal(err)
                }

                b, _ := json.MarshalIndent(res.GetResults(), "", " ")
                glg.Infof("%d - Results : %s\n\n", i+1, string(b))
                time.Sleep(1 * time.Second)
            }
            ```

            </details>

    1. Remove

        - Remove 400 indexed training datasets from the Vald agent.
            <details><summary>example code</summary><br>

            ```go
            for i := range ids [:insertCount] {
                _, err := client.Remove(ctx, &payload.Remove_Request{
                    Id: &payload.Object_ID{
                        Id: ids[i],
                    },
                })
                if err != nil {
                    glg.Fatal(err)
                }
                if i%10 == 0 {
                    glg.Infof("Removed %d", i)
                }
            }
            ```

            </details>
    </details>

## Cleanup

In the last, you can remove the deployed Vald Cluster by executing below command.

```bash
helm uninstall vald
```

## Next Steps

> Conglatulation! You completely entered into the Vald World!

Reviewer: ```suggestion
Congratulation! You completely entered the Vald World!
```

</details>

> ## Prepare the kubernetes cluster

Reviewer: ```suggestion
## Prepare the Kubernetes cluster
```

> 1. Prepare kubernetes cluster

Reviewer: ```suggestion
1. Prepare Kubernetes cluster
```

>     To complete get started, the kubernetes cluster is required.<br>

Reviewer: ```suggestion
    To complete get started, the Kubernetes cluster is required.<br>
```

>     Vald will run on Cloud Service such as GKE, AWS.

Reviewer: ```suggestion
    Vald will run on Cloud Service such as GKE, AWS.<br>
```

>     In the sense of trying to "Get-Started", [k3d](https://k3d.io/) or [kind](https://kind.sigs.k8s.io/) are easy kubernetes tools to use.

Reviewer: ```suggestion
    In the sense of trying to "Get-Started", [k3d](https://k3d.io/) or [kind](https://kind.sigs.k8s.io/) are easy to create Kubernetes cluster in your local environment.
```

> 1. Apply kubenetes metrics server

Reviewer: ```suggestion
1. Apply Kubernetes metrics server
```

    When finish deploy, you can check the Vald's pods status following command.

    ```bash
    kubectl get pods
    ```
    <details><summary>Example output</summary><br>
    If the deployment is successful, all Vald components should be running.

    ```bash
    NAME                                       READY   STATUS      RESTARTS   AGE
    vald-agent-ngt-0                           1/1     Running     0          7m12s
    vald-agent-ngt-1                           1/1     Running     0          7m12s
    vald-agent-ngt-2                           1/1     Running     0          7m12s
    vald-agent-ngt-3                           1/1     Running     0          7m12s
    vald-agent-ngt-4                           1/1     Running     0          7m12s
    vald-discoverer-7f9f697dbb-q44qh           1/1     Running     0          7m11s
    vald-lb-gateway-6b7b9f6948-4z5md           1/1     Running     0          7m12s
    vald-lb-gateway-6b7b9f6948-68g94           1/1     Running     0          6m56s
    vald-lb-gateway-6b7b9f6948-cvspq           1/1     Running     0          6m56s
    vald-manager-index-74c7b5ddd6-jrnlw        1/1     Running     0          7m12s
    ```

> ## Running Exapmle Code

Reviewer: ```suggestion
## Run Example Code
```

> In this chaptor, you will execute insert vectors, search vectors, and delete vectors to your Vald cluster using the exmaple code.<br>

Reviewer: ```suggestion
In this chapter, you will execute insert vectors, search vectors, and delete vectors to your Vald cluster using the example code.<br>
```

The [fashion-mnist](https://github.com/zalandoresearch/fashion-mnist) is used as a dataset for indexing and search query.

The example code is implemented Go and using [vald-client-go](https://github.com/vdaas/vald-client-go), one of the official Vald client library,  for requesting to Vald cluster.
Vald provides multiple language client libraries such as Go, Java, Node.js, Python, and so on.
If you are interested in, please refer to [SDKs](../../user-guides/sdks).<br>

1. Port Forward

    At first, port-foward is required to make request from your local environment possible.

    ```bash
    kubectl port-forward deployment/vald-lb-gateway 8081:8081
    ```

1. Download dataset

    Download [fashion-mnist](https://github.com/zalandoresearch/fashion-mnist) that is used as dataset for indexing and search query.

    ```bash
    # move to working directory
    cd example/client

    # download fashion-mnist testing dataset
    wget http://ann-benchmarks.com/fashion-mnist-784-euclidean.hdf5
    ```

> 1. Run

Reviewer: Maybe it is easier to understand 🤔
```suggestion
1. Run example
```

    We use [`example/client/main.go`](https://github.com/vdaas/vald/blob/master/example/client/main.go) to run the example.<br>

>     This example will insert and index 400 vectors into the Vald from the fashion-mnist dataset via grpc.

Reviewer: ```suggestion
    This example will insert and index 400 vectors to the Vald cluster from the fashion-mnist dataset via gRPC.<br>
```

>     And then after waiting for indexing, it will request for searching nearest vector at 10 times.

Reviewer: ```suggestion
    And then after waiting for indexing, it will request for searching nearest vector at 10 times.<br>
```

>     We use [`example/client/main.go`](https://github.com/vdaas/vald/blob/master/example/client/main.go) to run the example.<br>
>     This example will insert and index 400 vectors into the Vald from the fashion-mnist dataset via grpc.
>     And then after waiting for indexing, it will request for searching nearest vector at 10 times.
>     You will get the 10 nearest neighbor vectors for each search query.<br>
>     Run example codes by executing below command.

Reviewer 2: ```suggestion
    We use [`example/client/main.go`](https://github.com/vdaas/vald/blob/master/example/client/main.go) to run the example.<br>
    This example will insert and index 400 vectors into the Vald from the fashion-mnist dataset via gRPC.
    And then after waiting for indexing, it will request for searching the nearest vector 10 times.
    You will get the 10 nearest neighbor vectors for each search query.<br>
    Run example codes by executing below command.
```

    ```bash
    # run example
    go run main.go
    ```

>     The datailed explanation of exapmle code is below.

Reviewer: ```suggestion
    The detailed explanation of example code is below.
```

                    Config: &payload.Search_Config{
                        Num: 10,
                        Radius: -1,
                        Epsilon: 0.01,
                    }
                }
                if err != nil {
                    glg.Fatal(err)
                }

                b, _ := json.MarshalIndent(res.GetResults(), "", " ")
                glg.Infof("%d - Results : %s\n\n", i+1, string(b))
                time.Sleep(1 * time.Second)
            }
            ```

            </details>

    1. Remove

        - Remove 400 indexed training datasets from the Vald agent.
            <details><summary>example code</summary><br>

            ```go
            for i := range ids [:insertCount] {
                _, err := client.Remove(ctx, &payload.Remove_Request{
                    Id: &payload.Object_ID{
                        Id: ids[i],
                    },
                })
                if err != nil {
                    glg.Fatal(err)
                }
                if i%10 == 0 {
                    glg.Infof("Removed %d", i)
                }
            }
            ```

            </details>
    </details>

## Cleanup

> In the last, you can remove the all deployed Vald pods by executing below command.

Reviewer: ```suggestion
In last, you can remove the deployed Vald cluster by executing below command.
```

```bash
helm uninstall vald
```

## Next Steps

Conglatulation! You completely entered into the Vald World!

If you want, you can try other tutorials such as:
- [Vald Agent Standalone on k8s](../tutorial/vald-agent-standalone-on-k8s.md)

> - [Vald Agent on Docker](../tutorial/vald-agent-on-docker.md)

Reviewer: ```suggestion
- [Vald Agent on Docker](../tutorial/vald-agent-standalone-on-docker.md)
```
